### DISCUSSION TITLE
articles/2024/trying-pdm/

### DISCUSSION THREAD

**Discussion by giscus** [loopwerk/loopwerk.io#17] [Article discussions]
# articles/2024/trying-pdm/

After comparing uv to Poetry, I am trying out PDM. On paper it combines all the best things of Poetry and uv, without their downsides. How does it hold up?

https://www.loopwerk.io/articles/2024/trying-pdm/

**Comment by schneebuzz** [comment]
Thanks for the write up! It really helps as I am currently evaluating the use of PDM vs Poetry.

One thing I was wondering about, you state:

> Poetry does this a hundred times better by its add command not immediately installing packages.

This seems not true, the `add` command, as documented, *Adds a new dependency to pyproject.toml and installs it.*.
Am I missing something here?

Also, did you compare the editable installs feature of poetry vs. PDM and if so, what are your thoughts?
We are currently using a lot of "in-project-packages" that we install as editable. In PDM, those must go into a "dev" dependency group as PDM does not allow editable installs of prod dependencies.

**Reply by kevinrenskers** [reply]
Which version of Poetry are you running? I'm running 1.8.3, which is the latest.

And my `pyproject.toml` file looks like this;

```
[tool.poetry]
name = "my-project-name"
version = "0.1.0"
description = ""
readme = "readme.md"
package-mode = false

[tool.poetry.dependencies]
python = "^3.12"
Django = "^5.0.1"

[tool.poetry.group.dev]
optional = true

[tool.poetry.group.dev.dependencies]
ruff = "^0.6.0"

[tool.poetry.group.prod]
optional = true

[tool.poetry.group.prod.dependencies]
gunicorn = "^23.0.0"

[build-system]
requires = ["poetry-core"]
build-backend = "poetry.core.masonry.api"
```

I'm guessing your dev group is set up differently.

**Reply by schneebuzz** [reply]
Oh, I completely overlooked the fact that dependency groups are not `optional` by default. I am also using the latest versions for both PDM and poetry.

FYI: I found that `pdm add` has the `--no-sync` option, which makes PDM only modify the toml without installing the package.

Would you mind sharing the pyproject.toml for the PDM example as I am not able to reproduce the failing removal of dependencies. Anyway PDM also does not claim that it removes any dependency in my setup when installing only a specific group.

**Reply by kevinrenskers** [reply]
> FYI: I found that pdm add has the --no-sync option, which makes PDM only modify the toml without installing the package.

Oh, that's great! That would solve one of the major downsides I had with PDM. I'll give this a go and update the article if this indeed solves it.

> Would you mind sharing the pyproject.toml for the PDM example

```
[project]
name = "pdm-test"
version = "0.1.0"
description = "Default template for PDM package"
dependencies = [
    "django>=5.1.1",
]
requires-python = "==3.12.*"
readme = "README.md"
license = {text = "MIT"}

[project.optional-dependencies]
dev = [
    "ruff>=0.6.9",
]
prod = [
    "gunicorn>=23.0.0",
]

[tool.pdm]
distribution = false
```

**Comment by aureliojargas** [comment]
Nice writeup! This article and the "Poetry versus uv". Being a dissatisfied Poetry user myself, I'm having similar doubts on what tool to use.

Since `uv` is in a fast development cycle, I think your two concerns from the Conclusion were already addressed.

> And when it comes to handling optional dependency groups:
> - uv doesn’t have the concept of multiple dependency groups at all

In 2024-10-25 `uv` have added support for dependency groups (following PEP 735): https://github.com/astral-sh/uv/releases/tag/0.4.27 

> Finding and updating outdated packages:
> - uv fails by not making it possible to see outdated packages

This was just merged 4 days ago:

- [Add support for `pip list --outdated`](https://github.com/astral-sh/uv/pull/8872)
- [Add `uv tree --outdated`](https://github.com/astral-sh/uv/pull/8893)

And released 3 days ago in https://github.com/astral-sh/uv/releases/tag/0.5.0

**Reply by kevinrenskers** [reply]
Ah, thanks for the heads-up! That's great to hear, I'll definitely try that out and update the article if necessary.

**Reply by kevinrenskers** [reply]
All my biggest gripes have indeed been solved and I will definitely move my projects from Poetry to uv. New article: https://www.loopwerk.io/articles/2024/python-uv-revisited/.

**Comment by frostming** [comment]
Regarding the issue of selecting the Python interpreter for PDM, I am using version 2.21.0 and cannot reproduce. It's possible that the bug has already been fixed?

**Reply by kevinrenskers** [reply]
Are you using uv as the resolver? I only had the problem when that was the case. But yeah maybe it's solved now, no idea.

**Reply by frostming** [reply]
Yes using uv

**Reply by kevinrenskers** [reply]
I used whatever PDM version was the latest version at the time of writing this article, so October 4th. Next time I write an article like this I'll note the versions used.
